packages/feedback/src/screenshot/components/ScreenshotEditor.tsx

```diff
@@ -2,171 +2,372 @@ import type { FeedbackInternalOptions, FeedbackModalIntegration } from '@sentry/
-      const canvas = canvasRef.current;
-      if (!canvas) {
+    // Draw the screenshot into the background
+    hooks.useEffect(() => {
+      setCanvasSize(backgroundRef.current, screenshot.dpi);
+      paintImage(backgroundRef.current, screenshot.canvas);
+    }, [screenshot]);
+
+    // Draw the commands into the foreground
+    hooks.useEffect(() => {
+      setCanvasSize(foregroundRef.current, screenshot.dpi);
+      with2dContext(foregroundRef.current, { alpha: true }, (canvas, ctx) => {
+        ctx.clearRect(0, 0, canvas.width, canvas.height);
+      });
+      paintForeground(foregroundRef.current, strokeColor, drawCommands, undefined);
+    }, [drawCommands, strokeColor]);
+
+    // Draw into the output outputBuffer
+    hooks.useEffect(() => {
+      setCanvasSize(outputBuffer, screenshot.dpi);
+      paintImage(outputBuffer, screenshot.canvas);
+      with2dContext(DOCUMENT.createElement('canvas'), {alpha: true}, (foreground, ctx) => {
+        ctx.scale(screenshot.dpi, screenshot.dpi); // The scale needs to be set before we set the width/height and paint
+        foreground.width = screenshot.canvas.width;
+        foreground.height = screenshot.canvas.height;
+        paintForeground(foreground, strokeColor, drawCommands, undefined);
+        paintImage(outputBuffer, foreground);
+      });
+    }, [drawCommands, screenshot, strokeColor]);
+
+    const handleMouseDown = (e: MouseEvent): void => {
+      if (!action || !mouseRef.current) {
         return;
       }
 
-      canvas.width = imageDimensions.width * DPI;
-      canvas.height = imageDimensions.height * DPI;
-      canvas.style.width = `${imageDimensions.width}px`;
-      canvas.style.height = `${imageDimensions.height}px`;
-      const ctx = canvas.getContext('2d');
-      if (ctx) {
-        ctx.scale(DPI, DPI);
+      const boundingRect = mouseRef.current.getBoundingClientRect();
+      const startingPoint: DrawCommand = {
+        type: action,
+        x: e.offsetX / scaleFactor,
+        y: e.offsetY / scaleFactor,
+        w: 0,
+        h: 0,
+      };
+
+      const getDrawCommand = (startingPoint: DrawCommand, e: MouseEvent): DrawCommand => {
+        const x = (e.clientX - boundingRect.x) / scaleFactor;
+        const y = (e.clientY - boundingRect.y) / scaleFactor;
+        return {
+          type: startingPoint.type,
+          x: Math.min(startingPoint.x, x),
+          y: Math.min(startingPoint.y, y),
+          w: Math.abs(x - startingPoint.x),
+          h: Math.abs(y - startingPoint.y),
+        } as DrawCommand;
       }
-    }
 
-    function resize(): void {
-      const imageDimensions = getContainedSize(imageBuffer);
+      const handleMouseMove = (e: MouseEvent): void => {
+        with2dContext(foregroundRef.current, { alpha: true }, (canvas, ctx) => {
+          ctx.clearRect(0, 0, canvas.width, canvas.height);
+        });
+        paintForeground(foregroundRef.current, strokeColor, drawCommands, getDrawCommand(startingPoint, e));
```

Comment:
Any perf concerns here? Doesn't mousemove get fired quite frequently?

Reply:
yeah, mousemove is often. but we've got to redraw frequently for smooth pointer tracking. I believe this is the minimum stuff to draw:
- the foreground is separate from the background, so we're not actually re-drawing the original screenshot each frame
- what we are doing is clearing the canvas
- draw in the grey transparent background
  - gotta have that clear otherwise the background gets darker and darker
- for each drawCommand, we draw in the square, then if it's a highlight we'll crop out the interior where the grey transparency was. It's one fewer draw command from before.